plugins/fabric8-end2end-flow/src/main/java/com/redhat/che/end2end/End2EndFlowService.java

```diff
@@ -0,0 +1,217 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+@Path("fabric8-end2end")
+public class End2EndFlowService extends Service {
+  private static final Logger LOG = LoggerFactory.getLogger(End2EndFlowService.class);
+
+  private final String reCaptchaSiteKey;
+  private final String reCaptchaSecretKey;
+
+  private final Map<String, Function<String, String>> staticFilesFilters;
+
+  @Inject
+  public End2EndFlowService(
+      @Nullable @Named("che.fabric8.end2end.protect.site_key") String siteKey,
+      @Nullable @Named("che.fabric8.end2end.protect.secret_key") String secretKey) {
+    this.reCaptchaSiteKey = siteKey;
+    this.reCaptchaSecretKey = secretKey;
+    this.staticFilesFilters = new HashMap<>();
+    String siteKeyDecl;
+    if (siteKey != null && !siteKey.isEmpty()) {
+      siteKeyDecl = "var siteKey='" + siteKey + "';";
+    } else {
+      siteKeyDecl = "var siteKey='';";
+      LOG.warn("No ReCaptcha site key was provided. ReCaptcha user verification is disabled !");
+    }
+    staticFilesFilters.put(
+        "files/provision.html", line -> line.replace("const siteKey;", siteKeyDecl));
+  }
+
+  @GET
+  @Path("site-key")
+  public String siteKey() {
+    return reCaptchaSiteKey == null ? "" : reCaptchaSiteKey;
+  }
+
+  @POST
+  @Path("verify")
+  @Consumes("text/plain")
+  @Produces(APPLICATION_JSON)
+  public Response verifyReCaptchaToken(String token) {
+    int responseCode = 500;
+    try {
+      URL obj = new URL("https://www.google.com/recaptcha/api/siteverify");
+      HttpsURLConnection con = (HttpsURLConnection) obj.openConnection();
+
+      con.setRequestMethod("POST");
+
+      String urlParameters = "secret=" + reCaptchaSecretKey + "&response=" + token;
+
+      con.setDoOutput(true);
+      try (DataOutputStream wr = new DataOutputStream(con.getOutputStream())) {
+        wr.writeBytes(urlParameters);
+        wr.flush();
+      }
+
+      responseCode = con.getResponseCode();
+
+      if (responseCode == 200) {
+        StringBuffer response = new StringBuffer();
+        try (BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()))) {
+          String inputLine;
+          while ((inputLine = in.readLine()) != null) {
+            response.append(inputLine);
+          }
+        }
+        return Response.ok(response.toString()).build();
+      } else {
+        LOG.error(
+            "reCaptcha verification failed with the fllowing response code: {}", responseCode);
+      }
+    } catch (IOException e) {
+      LOG.error("Captcha verification failed with an exception", e);
+    }
+
+    return Response.status(responseCode).build();
+  }
+
+  private String getLog(String message, HttpServletRequest servletRequest) {
+    String ip = servletRequest.getHeader("X-Forwarded-For");
+    if (ip == null) {
+      ip = servletRequest.getRemoteAddr();
+    }
+
+    return new StringBuffer("[E2E Registration Flow - IP = ")
```

Comment:
We have 124k events like that on prod already. Can you elaborate on why we need to log all these events and what profits of having them compared to not having?

Reply:
This seems important to track the registration errors.
This huge number is obviously due to bots aggressively and constantly trying to register.
It might also indicate some erroneous code flow in the end2end automatic provisioning, which is important to track as well.

And even without bots / flow errors to track, it seems critical to be able to track the cases when true users didn't succeed the `che.openshift.io` registration fully.

Reply:
I doubt that we will analyze 124k log entries (and this is only for 2 weeks), but if you believe we need them let's leave them here.